### ISSUE TITLE
for { select { break } } gives unreachable code

### ISSUE THREAD

**Issue by orange4glace** [golang/lint#311]
```
for {
	select {
	case <-a:
		break
	case <-b:
		break
	}
	// golint says it's unreachable
	fmt.Println("It is reachable")
}
```

**Comment by dominikh** [comment]
golint does not look for or detect unreachable code; go vet does. And go vet does not flag the code you presented.

Please provide a full shell session with inputs and outputs that demonstrates the problem.

**Comment by EVODelavega** [comment]
I got the same output, but it's not a bug. `select` seems to behave like `switch`, a break doesn't affect the outer `for` loop. The solution to this would be to either:

* use a label for your loop and use a label break
* return instead of break, using either a defer or just moving the cleanup stuff to the case.

My specific use case was a select with a context (`case <-ctx.Done():`) and a buffered channel. I wanted to `for range` over the channel after the context got cancelled. Simply moving the `range` to the `Done` case made sense.
If you want to verify what the behaviour of the `break` you're using is, run this code:

```go
package main

import (
	"context"
	"fmt"
	"time"
)

func main() {
	ch := make(chan int, 10)
	ctx, cfunc := context.WithCancel(context.Background())
	go generate(ch)
	go routine(ctx, ch)
	time.Sleep(1 * time.Second)
	fmt.Println("calling cancel func")
	ch <- 123
	cfunc()
	ch <- 456
	fmt.Println("cancel called")
	close(ch)
	fmt.Println("done")
}

func generate(ch chan<- int) {
	for i := 0; i < 10; i++ {
		ch <- i
	}
}

func routine(ctx context.Context, ch <-chan int) {
	for {
		select {
		case i := <-ch:
			fmt.Println("Read from ch ", i)
		case <-ctx.Done():
			fmt.Println("ctx cancelled")
			break
		}
	}
	fmt.Println("end of loop")
	for i := range ch {
		fmt.Println("Read from ch ", i)
	}
}
```

Add a `default` case if you want to, and move the `sleep` around (eg sleep after canceling the context) and you'll note that _"ctx cancelled"_ shows up over and over again in the output.

**Comment by sunilvaltix** [comment]
I hit this issue too. I was trying breaking out of a for loop around select where if a gRPC stream is broken, I retry to connect. But my retry logic is never reached because its stuck in the inner for loop around select.

**Comment by sunilvaltix** [comment]
> "select seems to behave like switch, a break doesn't affect the outer for loop. "

Is this documented somewhere?

**Comment by griesemer** [comment]
@sunilvaltix https://tip.golang.org/ref/spec#Break_statements clearly states that a `break` statement may be used in a `select` statement.

**Comment by sunilvaltix** [comment]
@griesemer : Awesome! Thanks!
